water/lib/Dialect/Wave/IR/WaveOps.cpp

```diff
@@ -292,6 +321,57 @@ llvm::FailureOr<mlir::ChangeResult> wave::MmaOp::propagateBackward(
                                            "result", "accumulator", errs);
 }
 
+// Get the MMA result elements per thread for a given MMA kind
+static unsigned getMmaResultElementsPerThreadForMmaKind(mlir::MLIRContext *context, wave::WaveMmaKind kind) {
+  // Get the MMA specification (M, N, K dimensions and element types)
+  wave::WaveMmaSpec spec = wave::WaveMmaKindAttr::getSpec(context, kind);
+
+  // Elements per thread = (M × N) / threads_per_wave
+  // AMD GPU waves have 64 threads
+  constexpr unsigned threadsPerWave = 64;
+  unsigned totalElements = spec.m * spec.n;
+  return totalElements / threadsPerWave;
+}
+
+llvm::FailureOr<mlir::ChangeResult> wave::MmaOp::propagateElementsPerThreadForward(
+    llvm::ArrayRef<wave::ElementsPerThreadLatticeValue> operandElements,
+    llvm::MutableArrayRef<wave::ElementsPerThreadLatticeValue> resultElements,
+    llvm::raw_ostream &errs) {
+  // For MMA, the result elements per thread is determined by the MMA kind, not the operands
+  unsigned expectedElementsPerThread = getMmaResultElementsPerThreadForMmaKind(getContext(), getKind());
+  wave::ElementsPerThreadLatticeValue expectedResult(expectedElementsPerThread);
+
+  // Propagate to result
+  auto joined = wave::ElementsPerThreadLatticeValue::join(expectedResult, resultElements[0]);
+  if (joined.isTop() && !expectedResult.isTop() && !resultElements[0].isTop()) {
+    errs << "mismatched elements per thread for MMA result: expected " << expectedElementsPerThread
+         << " elements per thread for MMA kind " << getKind() << " but got (";
+    resultElements[0].print(errs);
+    errs << ")";
+    return mlir::failure();
+  }
+
+  if (joined != resultElements[0]) {
+    resultElements[0] = joined;
+    return mlir::ChangeResult::Change;
+  }
+
+  return mlir::ChangeResult::NoChange;
+}
+
+llvm::FailureOr<mlir::ChangeResult> wave::MmaOp::propagateElementsPerThreadBackward(
+    llvm::MutableArrayRef<wave::ElementsPerThreadLatticeValue> operandElements,
+    llvm::ArrayRef<wave::ElementsPerThreadLatticeValue> resultElements,
+    llvm::raw_ostream &errs) {
+  // For MMA, operands and result elements per thread may be different
+  // The result is determined by the MMA kind, operands can have their own values
+  // We don't propagate backwards for MMA since operands and result have independent constraints
```

Comment:
Shouldn't operands have the elements per thread value that is determined from the kind similarly to how results do? We don't propagate from _results_, but we still want to detect conflicts if they have mismatching values.